hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/specifications/TargetTypeSpecification.java

```diff
@@ -0,0 +1,87 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.repository.jpa.specifications;
+
+import org.eclipse.hawkbit.repository.jpa.model.JpaDistributionSetType;
+import org.eclipse.hawkbit.repository.jpa.model.JpaDistributionSetType_;
+import org.eclipse.hawkbit.repository.jpa.model.JpaTargetType;
+import org.eclipse.hawkbit.repository.jpa.model.JpaTargetType_;
+import org.eclipse.hawkbit.repository.model.DistributionSetType;
+import org.eclipse.hawkbit.repository.model.TargetType;
+import org.springframework.data.jpa.domain.Specification;
+
+import javax.persistence.criteria.SetJoin;
+import java.util.Collection;
+
+/**
+ * Specifications class for {@link TargetType}s. The class provides Spring Data JPQL
+ * Specifications.
+ *
+ */
+public final class TargetTypeSpecification {
+
+    private TargetTypeSpecification() {
+        // utility class
+    }
+
+    /**
+     * {@link Specification} for retrieving {@link TargetType}s by controllerId
+     *
+     * @param id
+     *            to search for
+     *
+     * @return the {@link TargetType} {@link Specification}
+     */
+    public static Specification<JpaTargetType> hasId(final Long id) {
+        return (targetRoot, query, cb) -> cb.equal(targetRoot.get(JpaTargetType_.id), id);
+    }
+
+    /**
+     * {@link Specification} for retrieving {@link TargetType}s by controllerId
+     *
+     * @param ids
+     *            to search for
+     *
+     * @return the {@link TargetType} {@link Specification}
+     */
+    public static Specification<JpaTargetType> hasIdIn(final Collection<Long> ids) {
+        return (targetRoot, query, cb) -> targetRoot.get(JpaTargetType_.id).in(ids);
+    }
+
+    /**
+     * {@link Specification} for retrieving {@link TargetType}s based on a
+     * {@link DistributionSetType} name.
+     *
+     * @param dsTypeId
+     *            to search for
+     *
+     * @return the {@link TargetType} {@link Specification}
+     */
+    public static Specification<JpaTargetType> hasDsSetType(final Long dsTypeId) {
+        return (targetRoot, query, cb) -> {
+            final SetJoin<JpaTargetType, JpaDistributionSetType> join = targetRoot.join(JpaTargetType_.distributionSetTypes);
+            return cb.equal(join.get(JpaDistributionSetType_.id), dsTypeId);
+        };
+    }
+
+
+    /**
+     * {@link Specification} for retrieving {@link TargetType} with
+     * given {@link TargetType#getName()} including fetching the
+     * elements list.
+     *
+     * @param name
+     *            to search
+     * @return the {@link TargetType} {@link Specification}
+     */
+    public static Specification<JpaTargetType> byName(final String name) {
+        return (targetRoot, query, cb) -> cb.equal(targetRoot.get(JpaTargetType_.name), name);
+    }
+
```

Comment:
Please also add specifications to get `TargetType`s by assigned `Target`s. As we use both the `long id` and `String controllerId` to identify targets and searching should be possible for collections as well as single entities I suggest the following four entries:

- hasTarget(long id)
- hasTargetIn(Collection\<ids\> ids)
- hasTargetControllerId(String controllerId)
- hasTargetControllerIdIn (Collection\<String\> controllerIds)

The specs could look something like this:
```java
final SetJoin<JpaTargetType, JpaTarget> join = targetRoot.join(JpaTargetType_.targets);
return join.get(JpaTarget_.controllerId).in(controllerIds);
```